File: javascript/src/context.js
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2025 Ian Thomas

import { ExpressionParser } from '../lib/expression-parser/expressionParser.js';

const expressionParser = new ExpressionParser();

// Use the ExpressionParser to evaluate an expression
// Unless it's a simple bool or number, in which case just use that.
// dump_eval will fill an array with evaluation debug steps
export function evalExpression(val, context, dump_eval = null) {
  if (typeof val==="boolean"||typeof val==="number")
    return val;

  const expression = expressionParser.parse(val);
  return expression.evaluate(context, dump_eval);
}

// Evaluate and copy any new entries in 'properties' into the context
// dump_eval will fill an array with evaluation debug steps
export function initContext(context, properties, dump_eval = null) {
  for (const [propName, expression] of Object.entries(properties)) {
    if (dump_eval)
      dump_eval.push(`InitContext: Evaluating ${propName} = ${expression}`);
    const result = evalExpression(expression, context, dump_eval);
    if (propName in context)
      throw new Error(`Trying to initialise propert '${propName}' in context when it already exists.`)
    context[propName] = result;
  }
}

// Evaluate any entries in updates and apply them to the context, but they must already exist
// dump_eval will fill an array with evaluation debug steps
export function updateContext(context, updates, dump_eval = null) {
  for (const [propName, expression] of Object.entries(updates)) {
    if (dump_eval)
      dump_eval.push(`UpdateContext: Evaluating ${propName} = ${expression}`);
    const result = evalExpression(expression, context, dump_eval);
    //console.log(`Setting ${varName} to ${result}`);
    if (!(propName in context)) {
      throw new Error(`Context var: '${propName}' undefined.`);
    }
    context[propName] = result;
  }
}

// For debugging
export function dumpContext(context) {
  let out = [];
  for (const [propName, expression] of Object.entries(context)) {
    if (typeof expression === 'function') {
      out.push(`${propName} = <function>`);
    }
    else if (typeof expression==="boolean"||typeof expression==="number"||typeof expression==="string") {
      const result = evalExpression(expression, context);
      out.push(`${propName} = ${result}`);
    }
  }
  return out.join('\n');
}

File: cpp/src/context.cpp
/*
 * This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
 * Copyright (c) 2025 Ian Thomas
 */

 #include "storylet_framework/context.h"
 
 #include <iostream>
 #include <string>
 #include <stdexcept>
 #include <sstream>

namespace StoryletFramework
{

    ExpressionParser::Parser expressionParser;

    // Evaluate an expression
    std::any ContextUtils::EvalExpression(const std::any& val, const Context& context, DumpEval* dumpEval)
    {
        if (val.type() == typeid(bool) || val.type() == typeid(double) || val.type() == typeid(int))
        {
            return val;
        }

        if (val.type() == typeid(std::string))
        {
            const auto& str = std::any_cast<std::string>(val);
            auto expression = expressionParser.Parse(str);
            if (!expression)
            {
                throw std::invalid_argument("Expression result should never be null.");
            }
            return expression->Evaluate(context, dumpEval);
        }

        throw std::invalid_argument("Expression text cannot be null or empty.");
    }

    // Initialize context with properties
    void ContextUtils::InitContext(Context& context, const KeyedMap& properties, DumpEval* dumpEval)
    {
        for (const auto& kvp : properties)
        {
            const auto& propName = kvp.first;
            const auto& expression = kvp.second;

            if (context.find(propName) != context.end())
            {
                throw std::invalid_argument("Trying to initialize property '" + propName + "' in context when it already exists.");
            }

            if (dumpEval)
            {
                dumpEval->push_back("InitContext: Evaluating " + propName + " = " + std::any_cast<std::string>(expression));
            }

            auto result = EvalExpression(expression, context, dumpEval);
            context[propName] = result;
        }
    }

    // Update context with updates
    void ContextUtils::UpdateContext(Context& context, const KeyedMap& updates, DumpEval* dumpEval)
    {
        for (const auto& kvp : updates)
        {
            const auto& propName = kvp.first;
            const auto& expression = kvp.second;

            if (context.find(propName) == context.end())
            {
                throw std::out_of_range("Context variable '" + propName + "' is undefined.");
            }

            if (dumpEval)
            {
                dumpEval->push_back("UpdateContext: Evaluating " + propName + " = " + std::any_cast<std::string>(expression));
            }

            auto result = EvalExpression(expression, context, dumpEval);

            if (dumpEval)
            {
                dumpEval->push_back("Setting " + propName + " to " + std::any_cast<std::string>(expression));
            }

            context[propName] = result;
        }
    }

    // Dump the context as a string for debugging
    std::string ContextUtils::DumpContext(const Context& context)
    {
        std::ostringstream output;

        for (const auto& kvp : context)
        {
            const auto& propName = kvp.first;
            const auto& expression = kvp.second;

            if (expression.type() == typeid(ExpressionParser::FunctionWrapper))
            {
                output << propName << " = <function>\n";
            }
            else if (expression.type() == typeid(bool))
            {
                output << propName << " = " << (std::any_cast<bool>(expression) ? "true" : "false") << "\n";
            }
            else if (expression.type() == typeid(int))
            {
                output << propName << " = " << std::any_cast<int>(expression) << "\n";
            } 
            else if (expression.type() == typeid(std::string))
            {
                output << propName << " = \"" << std::any_cast<std::string>(expression) << "\"\n";
            } 
            else if (expression.type() == typeid(double))
            {
                output << propName << " = " << std::any_cast<double>(expression) << "\n";
            } 
            else
            {
                output << propName << " = <unknown type>\n";
            }
        }

        return output.str();
    }
}